### PR TITLE
Jjiang/key metadata pair map

### PR DESCRIPTION
**Pull request by jerryjiangabc**
Originally, SmallArrayBasedLongToDoubleMap does not contain duplicate keys, as we only need to record the engagement types in it, and if one user replies or quote a tweet ten times, he/she will only appear in the social proof once. Now, these ten reply edges have different metadata, and GraphJet needs to return all of them as separate social proofs. 

This rb changes the dedup logic in SmallArrayBasedLongToDoubleMap from checking keys to checking key, metadata pairs. Along with this change, I add a new public function, uniqueKeysSize() which returns the number of unique keys in SmallArrayBasedLongToDoubleMap. This is needed because clients use the uniqueKeysSize to check the social proof threshold in the algorithms.

In summary, algorithms query uniqueKeysSize() to check against social proof threshold but return the whole array of keys and metadata when hydrating the social proof response.

Along with the above change, we will not apply maxNumSocialProofs cutoff inside TopSecondDegreeByCountForTweet, because it only makes sense to operate on the unique keys. So clients might start to see more social proofs returned from GraphJet than before, which I think is a plus.

### CODE REVIEW

File: graphjet-core/src/main/java/com/twitter/graphjet/hashing/SmallArrayBasedLongToDoubleMap.java [hunk 1]
```diff
@@ -22,35 +22,88 @@
...
+ * or not. When the size of the map is equal to ADD_KEYS_TO_SET_THRESHOLD, it adds all keys and
+ * metadata to a map and starts to use the map for dedupping.
  */
 public class SmallArrayBasedLongToDoubleMap {
+
+  /**
+   * Holder class of a pair of key and metadata, which is used during dedupping when the size of the
+   * map grows beyond ADD_KEYS_TO_SET_THRESHOLD.
+   */
+  private static final class Pair {
```
Comment (rossides) [inline graphjet-core/src/main/java/com/twitter/graphjet/hashing/SmallArrayBasedLongToDoubleMap.java]: Since this is internal and the naming is "not important" interface-wise, should we reuse the 3rd party apache pair class?

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/GeneratorHelper.java [hunk 2]
```diff
@@ -53,7 +52,8 @@ private GeneratorHelper() {
         if (socialProof.size() > 1) {
           socialProof.sort();
         }
-        socialProof.trim(maxSocialProofSize);
+
+        socialProof.trim(socialProof.size());
```
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/GeneratorHelper.java]: I want to get rid of this line too. Internally, the capacity of the array grows as a multiple of 2. There are situations like the size of the array is 5, but the capacity is 8. If we do not call trim function here, the following line, new LongArrayList(socialProof.keys()), will generate 3 entries with zero in them. So this line saves clients from trimming zeros on their own.

File: graphjet-core/src/main/java/com/twitter/graphjet/algorithms/salsa/SalsaSelectResults.java [hunk 3]
```diff
@@ -142,7 +140,7 @@ public SalsaResponse pickTopNodes() {
           socialProof.sort();
         }
 
-        socialProof.trim(maxSocialProofSize);
+        socialProof.trim(socialProof.size());
```
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/algorithms/salsa/SalsaSelectResults.java]: Ditto!

File: graphjet-core/src/main/java/com/twitter/graphjet/hashing/SmallArrayBasedLongToDoubleMap.java (hunk 1 above)
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/hashing/SmallArrayBasedLongToDoubleMap.java]: All 3rd party apache pair classes are based on generics, and when it holds a pair of longs, it will autobox them into object Longs, which this internal class avoids.

**Comment by jerryjiangabc** [comment]
>> Just bringing it up in case we need to update or inform the clients.
Yes, that is an excellent point. I plan to address this from scala service layer, before returning the response to clients.
